src/services/reference_services.py

```diff
@@ -13,11 +13,10 @@ def add_book(self, authors: str, title: str, year: int, publisher: str, publishe
         try:
 
             new_book = Book(authors, title, year, publisher, publisher_address)
-
-            # Search for a duplicate title in the database
-            # if self.get_book_by_title is not None:
-            #     print("Book already exists in the database")
-            #     raise Exception("Book already exists in the database) TODO
+            
+            # Check if the book with the same title already exists
+            if self.get_book_by_title(new_book.title):
+                return False
```

Comment:
Should raise Exception instead of direct False-return.

Reply:
I think the current test should work for this as well, no? Just add a self.mock_dataprocessing.assert_called_once() or assert_called_once_with('Garwin, Robert',
                                              'Clean Code: A Handbook of Agile Software Craftsmanship',
                                              2008, 'Prentice Hall', 'Bakerstreet 123')
after the self.assertFalse

Reply:
The current test indeed worked, despite removing the return False. I mixed some of the trial-error attempts I made.

However, I did not get 10/10 in the pylint tests with "raise Exception" so changed that into "raise ValueError" to get 10/10 in pylint tests. Hope that is OK. https://github.com/Wincewind/ohtu-s23-miniprojekti/blob/06ae2a2122d835546ab01825b96da5a4bb969fcf/src/services/reference_services.py#L20